runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/LockFreeHistogram.java

```diff
@@ -0,0 +1,216 @@
+   */
+  @AutoValue
+  public abstract static class Snapshot {
+    public abstract OutlierStatistic underflowStatistic();
+
+    public abstract OutlierStatistic overflowStatistic();
+
+    public abstract ImmutableLongArray buckets();
+
+    public abstract HistogramData.BucketType bucketType();
+
+    public static Snapshot create(
+        OutlierStatistic underflowStatistic,
+        OutlierStatistic overflowStatistic,
+        ImmutableLongArray buckets,
+        HistogramData.BucketType bucketType) {
+      return new AutoValue_LockFreeHistogram_Snapshot(
+          underflowStatistic, overflowStatistic, buckets, bucketType);
+    }
+
+    @Memoized
+    public long totalCount() {
+      long count = 0;
+      count += underflowStatistic().count();
+      count += overflowStatistic().count();
+      count += buckets().stream().sum();
+
+      return count;
+    }
+  }
+
+  /**
+   * Extract a delta update of this histogram. Update represents values that have been recorded in
+   * this histogram since the last time this method was called.
+   *
+   * <p>If this histogram is being updated concurrent to this method, then the returned snapshot is
+   * not guarenteed to contain those updates. However, those updates are not dropped and will be
+   * represented in a future call to this method.
+   *
+   * <p>If this histogram has not been updated since the last call to this method, an empty optional
+   * is returned.
+   */
+  public Optional<Snapshot> getSnapshotAndReset() {
+    if (!dirty.getAndSet(false)) {
+      return Optional.empty();
+    }
+
+    ImmutableLongArray.Builder bucketsSnapshotBuilder =
+        ImmutableLongArray.builder(buckets.length());
+    for (int i = 0; i < buckets.length(); i++) {
+      bucketsSnapshotBuilder.add(buckets.getAndSet(i, 0));
+    }
+    OutlierStatistic overflowSnapshot = overflowStatistic.getAndSet(OutlierStatistic.EMPTY);
+    OutlierStatistic underflowSnapshot = underflowStatistic.getAndSet(OutlierStatistic.EMPTY);
+
+    return Optional.of(
+        Snapshot.create(
+            underflowSnapshot, overflowSnapshot, bucketsSnapshotBuilder.build(), bucketType));
+  }
+
+  /** Create a histogram. */
+  public LockFreeHistogram(KV<MetricName, HistogramData.BucketType> kv) {
+    this.name = kv.getKey();
+    this.bucketType = kv.getValue();
+    this.buckets = new AtomicLongArray(bucketType.getNumBuckets());
+  }
+
+  @Override
+  public MetricName getName() {
+    return name;
+  }
+
+  private void updateInternal(double value) {
+    double rangeTo = bucketType.getRangeTo();
+    double rangeFrom = bucketType.getRangeFrom();
+    if (value >= rangeTo) {
+      recordTopRecordsValue(value);
+    } else if (value < rangeFrom) {
+      recordBottomRecordsValue(value);
+    } else {
+      recordInBoundsValue(value);
+    }
+  }
+
+  @Override
+  public void update(double value) {
+    updateInternal(value);
+    dirty.set(true);
```

Comment:
is this threadsafe?  seems racy since it is not using the atomics API or locks

Reply:
It won’t throw any ConcurrentModificationsErrors but, you’re right we don’t atomically flip this variable and update the metric. 

I’m using this as an optimization for the thread that extracts metrics if there’s a LockFreeHistogram that’s not being written into. However, this variable is not necessary from a correctness standpoint (see details below), so I can remove it if you’d like.

This variable is updated when recording metrics/extracting updates. It does not affect the record path. On the extract path, we can either get a false positive (`dirty` is `true` when there’s no updates in the histogram) or a false negative (`dirty` is `false` when there are updates to the underlying histogram). 

- In the false positive scenario, after we extract a snapshot we check that it’s non-empty before sending it to DFE so empty snapshots will just get dropped.
- In the false negative scenario, since we always set this value to true after recording a value in the histogram a future call to `getSnapshotAndReset` will see `dirty` set to `true` and extract the update.